### PR TITLE
Fork TLSConfig for each encrypted connection

### PR DESCRIPTION
**Pull request by sanyo0714**
Signed-off-by: sanyo <yeshengan.ysa@alibaba-inc.com>

### 1. Why is this pull request needed and what does it do?

Fork TLSConfig for each encrypted connection

### 2. Which issues (if any) are related?

#5703

### CODE REVIEW

**Comment by chrisohaver** [comment]
Can you add a test?

**Comment by sanyo0714** [comment]
> Can you add a test?


Can I add the test case this way?
```
caddy.Quiet = true
dnsserver.Quiet = true

corefile := `
https://.:8443 tls://.:8080 {
tls test_cert.pem test_key.pem test_ca.pem
}`
server, err := caddy.Start(caddy.CaddyfileInput{
    Contents:       []byte(corefile),
    Filepath:       "config",
    ServerTypeName: "dns",
})
if err != nil {
    t.Errorf("Failed to start server for input %q:\n%s", err, corefile)
}

serItemSet := make(map[string]struct{}, 2)
for _, s := range server.Servers() {
    sv := reflect.ValueOf(s)
    if sv.CanInterface() {
        ser := sv.FieldByName("server").Elem()
        switch ser.Type().String() {

        // ServerTLS format &{0xc00029e540 0xc00041ec00}
        // ServerHTTPS format &{0xc00029e5a0 0xc0004422a0 0xc00012b110 0xc00041ec00 0x170e6a0}
        // check that all items of ServerTLS differ from ServerHTTPS
        case "*dnsserver.ServerTLS", "*dnsserver.ServerHTTPS":
            ser := fmt.Sprintf("%v", ser)
            for _, item := range strings.Split(ser[2:len(ser)-1], " ") {
                if _, ok := serItemSet[item]; ok {
                    log.Fatalf("dnsserver.ServerTLS,dnsserver.ServerHTTPS has same item:[%s]", item)
                } else {
                    serItemSet[item] = struct{}{}
                }
            }
        }
    }
}
server.Stop()
```

**Comment by chrisohaver** [comment]
> > Can you add a test?
> 
> Can I add the test case this way?
> ...

That suggested test case would verify that your solution is in place, but doesn't really verify that the solution resolves the problem ("dot is broken").

A better test would be a test like the ones in `/test` that runs the server and sends it test queries that verify that dot functions as expected in the given configuration.

File: test/tls_test.go [hunk 1]
```diff
@@ -2,21 +2,23 @@ package test
 
 import (
 	"crypto/tls"
+	"fmt"
 	"testing"
 
 	"github.com/miekg/dns"
 )
 
 func TestDNSoverTLS(t *testing.T) {
```
Comment (chrisohaver) [inline test/tls_test.go]: Can you rename this test to `TestTLS`, and make it execute two test cases: the existing unaltered dot case, and the new dot+doh case?

**Review by chrisohaver — COMMENTED** [review]
Thanks for finding and fixing this. The test case looks good, one minor comment.

**Review by chrisohaver — APPROVED** [review]

**Pull request merged**